pkg/app/zipper/tldcache.go

```diff
@@ -13,16 +15,25 @@ func (app *App) probeTopLevelDomains(ms *PrometheusMetrics) {
 	probeTicker := time.NewTicker(time.Duration(app.Config.InternalRoutingCache) * time.Second) // TODO: The ticker resources are never freed
 	for {
 		topLevelDomainCache := make(map[string][]*backend.Backend)
-		ms.TLDCacheProbeReqTotal.Add(float64(len(app.Backends)))
-		for i := 0; i < len(app.Backends); i++ {
-			topLevelDomains, err := getTopLevelDomains(app.Backends[i])
-			if err != nil {
-				// this could add a lot of noise to logs
-				// lg.Error("failed to probe TLD cache for a backend", zap.Error(err), zap.String("backend", app.Backends[i].GetServerAddress()))
-				ms.TLDCacheProbeErrors.Inc()
-			}
-			for _, topLevelDomain := range topLevelDomains {
-				topLevelDomainCache[topLevelDomain] = append(topLevelDomainCache[topLevelDomain], &app.Backends[i])
+		// We should always have empty prefix to get default TLDs
+		allPrefixes := []string{""}
+		allPrefixes = append(allPrefixes, app.Config.TLDCacheExtraPrefixes...)
+		// Sorting to avoid involving unnecessary backends
+		sortedPrefixes := sortedByNsCount(allPrefixes)
+		for _, prefix := range sortedPrefixes {
+			prefix = trimPrefix(prefix)
+			bs := getBackendsForPrefix(prefix, app.Backends, topLevelDomainCache)
+			for i := 0; i < len(bs); i++ {
```

Comment:
nit: maybe we could use range() here? it's a tad simpler and is kinda default usually